### DISCUSSION TITLE
Del 2 - overkomplisert?

### DISCUSSION THREAD

**Discussion by jorgenhelgheim** [AmroAlnajar/FizzBuzzJazzFuzz#7] [General]
Navngivning er vanskelig. Det tror jeg alle er enige i. Jeg stoppet litt opp ved GetRuleModel, som jeg tenkte ikke trenger Model. det er vel en Rule bare. Men så brukes den bare til å returnere en model inne i IGameRuleSettings. Kunne ikke egenskapene (properties) i GetRuleModel bare blitt plassert rett i interfacet så slipper du GetRuleModel klassen helt. Og slipper å bekymre deg for å navngi enda en klasse/modell.

Når vi er inne på navngivning så ser jeg ikke helt poenget med å ha Enum som postfix på en enum klasse. Det er implisitt angitt av typen. Men akuratt det er litt småpirk.

**Comment by jorgenhelgheim** [comment]
Jeg ser gjerne over en gang til etter at du har sett og besvart disse to diskusjonene.

**Comment by AmroAlnajar** [comment]
Når en metode skal returnere flere verdier, så synes jeg at det er ryddigere å lage en Model som har alt samlet i ett sted. Tenker at det er også bedre for fremtiden da det kan komme nye krav som gjør at GetRuleSettings() skal returnere noe mer enn bare Range og Words, 

for meg, så ser `return new GameRuleModel { RuleRange = ruleRange, RuleWords = ruleWords };` bedre ut enn 
`return (ruleRange, ruleWords)`

en fordel med å returnere en Model fra GetRuleSettings() er at jeg kan bruke verdiene på en mer oversiktlig måte.

Eks:
ved å returnere en Model så kan jeg bruke metoden slikt: `var gameRules = _ruleSettings.GetRuleSettings().RuleWords.Select(x => new GameRules(x));`

uten å bruke en Model så vil det se noe slikt ut: `var gameRules = _ruleSettings.GetRuleSettings().Item1.Select(x => new GameRules(x));`

Merk på den siste linja, at Item1 kan være litt forvirrende. Derfor bruker jeg Models på nesten alt som returnerer mer enn en verdi .


Ifht Enums så er jeg Enig at vi kunne for eksempel returnere to string som "Fizz" og "3", og det kan være at en Enum her er en overkill, men her så unngår jeg magic strings, som jeg synes vil være bedre å holde organisert via å bruke  en Enum.

NB:
WordsEnum er et navn som jeg ikke er helt fornøyd med, fortsatt usikker på hva denne Enum'n burde bli kalt.

**Reply by jorgenhelgheim** [reply]
Enig at enum er bedre enn "magic strings", så det er et fint valg. Ja også usikekr på hva enum'en burde hete evt:). 

Vdr interface så var det slik jeg tenkte du kunne unngått bruken av GetRuleModel:
```
public interface IGameRuleSettings
{
    public IEnumerable<WordsEnum> RuleWords { get;  }
    public IEnumerable<int> RuleRange { get; }
}
```
Og bruken av den blir da
```
public class FizzBuzzRuleSettings : IGameRuleSettings
{
    public IEnumerable<WordsEnum> RuleWords => new List<WordsEnum> { WordsEnum.Fizz, WordsEnum.Buzz };
    public IEnumerable<int> RuleRange => Enumerable.Range(1, 100);
}
```
Med litt refakturere etter disse endringene, kan du slette GameRuleModel

**Reply by AmroAlnajar** [reply]
Sjekket koden og jeg ser hva du mente der. Jeg skulle ha tenkt på dette tidligere, er mye lettere enn å ha en klasse som skaper ett nytt objekt av GameRuleModel.
